src/core/vector/qgsvectorlayer.cpp

```diff
@@ -1007,14 +1015,19 @@ void QgsVectorLayer::updateDefaultValues( QgsFeatureId fid, QgsFeature feature )
 }
 
 QgsRectangle QgsVectorLayer::extent() const
+{
+  return extent3D().toRectangle();
```

Comment:
I'm wondering if it is the best strategy from a performance point of view. For example a provider, let's say the OGR one, might have a very inefficient extent3d() implementation that needs to iterate over all features (there's no OGR_L_GetExtent3D() in OGR currently, and even when there will be one, most formats, including GeoPackage, might have to fallback to a feature iteration strategy), whereas it can have a fast (2D) extent() one.

Reply:
I guess there's also a shortcut which the extent3d method can use when we know the layer geometry type doesn't have z values -- in that case extent3d could just return the (hopefully optimised) call to extent().

Reply:
Each implementation can  override this default. I will push the implementation/adjustment in next PRs.